### PR TITLE
Implement bundled skill file reads (1.3.4)

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch implements roadmap task `(1.3.4)` by adding a read-only `skill_read_file` interface for bundled skill resources. It gives the model a narrow, skill-scoped way to read `SKILL.md`, `references/**`, and allowed text assets from loaded bundles without exposing raw filesystem access.

Roadmap task: `(1.3.4)`
Execplan: [docs/execplans/1-3-4-skill-read-file-interface.md](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/docs/execplans/1-3-4-skill-read-file-interface.md)

## Review walkthrough

- Start with [src/skills/file_read.rs](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/skills/file_read.rs) for the model-facing response contract and policy entrypoint.
- Review [src/skills/file_read/io.rs](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/skills/file_read/io.rs) and [src/skills/file_read/validation.rs](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/skills/file_read/validation.rs) for path validation, symlink handling, size limits, and inline-text handling.
- Then review [src/tools/builtin/skill_tools.rs](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/tools/builtin/skill_tools.rs), [src/tools/registry/builtins.rs](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/tools/registry/builtins.rs), and [src/skills/attenuation.rs](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/skills/attenuation.rs) for tool registration and attenuation.
- Finish with [src/skills/file_read/tests.rs](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/skills/file_read/tests.rs), [src/tools/builtin/skill_tools/tests.rs](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/tools/builtin/skill_tools/tests.rs), and [src/tools/builtin/skill_tools/features/skill_read_file.feature](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/src/tools/builtin/skill_tools/features/skill_read_file.feature) for unit, property, schema, and `rstest-bdd` coverage.
- Documentation updates live in [docs/users-guide.md](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/docs/users-guide.md), [docs/agent-skills-support.md](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/docs/agent-skills-support.md), [docs/developers-guide.md](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/docs/developers-guide.md), [docs/roadmap.md](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/docs/roadmap.md), and [FEATURE_PARITY.md](https://github.com/leynos/axinite/blob/fbfd0f30baf326767ebd571bb68c8328363b7612/FEATURE_PARITY.md).

## Validation

- `cargo test --features test-helpers --lib skills::file_read -- --nocapture`: passed, 17 tests.
- `cargo test --features test-helpers --lib skill -- --nocapture`: passed, 270 tests.
- `bunx markdownlint-cli2 docs/execplans/1-3-4-skill-read-file-interface.md docs/users-guide.md docs/agent-skills-support.md docs/developers-guide.md docs/axinite-architecture-overview.md docs/roadmap.md FEATURE_PARITY.md`: passed with 0 errors.
- `git diff --check`: passed.
- `make check-fmt`: passed.
- `make lint`: passed.
- `make test`: passed, including 4091 nextest tests and 5 GitHub tool tests.
- `make all`: passed after the final source and ExecPlan updates, including 4091 nextest tests and 5 GitHub tool tests.
- `coderabbit review --agent`: actionable findings were addressed. The final confirmation retry is blocked by CodeRabbit service rate limits after the last non-Unix fallback comment; see the ExecPlan notes for the command logs.

## Notes

The non-Unix file-open fallback documents its current mitigation explicitly: it relies on the pre-open `symlink_metadata` rejection and the post-open size revalidation, with a TODO for platform-specific atomic no-follow semantics if non-Unix deployments become first-class for this tool.

No Python e2e test was added because this branch changes a model-facing built-in tool contract, not a new HTTP route, CLI command, persistence workflow, UI flow, or network boundary. The Rust adapter and `rstest-bdd` scenarios cover the externally visible tool behaviour.

## References

- Lody session: https://lody.ai/leynos/sessions/6c92fc0f-e404-4a10-ace1-30a36066de50

## Summary by Sourcery

Add a read-only `skill_read_file` tool and underlying skill-scoped file read subsystem so models can read specific bundled skill resources via bundle-relative paths without exposing host filesystem details.

New Features:
- Introduce the `skill_read_file` built-in tool that reads allowed text files from loaded skill bundles using a strict JSON schema.
- Add a `skills::file_read` module that enforces bundle-relative path policy, scoped filesystem I/O, and structured success/error responses for skill file reads.

Enhancements:
- Register `skill_read_file` in the tool registry and include it in the installed-skill read-only attenuation set.
- Extend documentation and architecture notes to describe bundled file read behaviour, path policy, and developer extension points.
- Update feature parity and roadmap docs to reflect completion of the bundled skill file read interface and scoped read capability.

Documentation:
- Add an ExecPlan documenting the design, constraints, risks, and validation steps for the `skill_read_file` interface.
- Update user and developer guides, architecture overview, agent skills support docs, and feature parity notes to describe the new skill-scoped file reading tool, its schema, and limitations.

Tests:
- Add unit, property-based, and behavioural (`rstest-bdd`) tests covering path validation, inline text reads, non-inline assets, error codes, and symlink/traversal rejection for skill bundle file reads.
- Extend skill tool tests to cover the new `skill_read_file` tool schema, execution behaviour, and BDD scenarios for allowed reads and denied filesystem traversal.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- In `read_validated_skill_file` / `resolve_and_validate_target` you canonicalize the skill root on every read; if `skill_read_file` is used frequently this could add avoidable filesystem overhead, so consider storing a precomputed canonical root on `LoadedSkill` (or similar) and reusing it here.
- The non-Unix `open_readonly_no_follow` fallback currently has weaker guarantees than the Unix path (no O_NOFOLLOW); given the explicit TODO, it might be worth also emitting a debug or warn log when this branch is used so operators are aware that symlink protections are reduced on those platforms.

## Individual Comments

### Comment 1
<location path="src/skills/file_read/tests.rs" line_range="50-16" />
<code_context>
+    SkillReadFixture { _dir: dir, skill }
+}
+
+#[rstest]
+#[tokio::test]
+async fn reads_bundle_reference_text(skill_read_fixture: SkillReadFixture) {
+    let response = read_skill_file(&skill_read_fixture.skill, "references/usage.md").await;
+
+    assert_eq!(
+        response,
+        SkillReadFileResponse::Success(SkillReadFileSuccess {
+            skill: "deploy-docs".to_string(),
+            path: "references/usage.md".to_string(),
+            mime_type: "text/markdown".to_string(),
+            content: "# Usage\n".to_string(),
+        })
+    );
+}
+
+#[rstest]
</code_context>
<issue_to_address>
**suggestion (testing):** Add a boundary test for exactly MAX_SKILL_READ_FILE_BYTES-sized files

The existing `oversized_text_returns_file_too_large` test covers the `> MAX_SKILL_READ_FILE_BYTES` path. Please add a test where the file size is exactly `MAX_SKILL_READ_FILE_BYTES` and assert it succeeds, so the boundary behavior (including the `usize` cast in `read_file_contents`) is explicitly covered and less likely to regress.

Suggested implementation:

```rust
    SkillReadFixture { _dir: dir, skill }
}

#[rstest]
#[tokio::test]
async fn reads_bundle_reference_text(skill_read_fixture: SkillReadFixture) {
    let response = read_skill_file(&skill_read_fixture.skill, "references/usage.md").await;

    assert_eq!(
        response,
        SkillReadFileResponse::Success(SkillReadFileSuccess {
            skill: "deploy-docs".to_string(),
            path: "references/usage.md".to_string(),
            mime_type: "text/markdown".to_string(),
            content: "# Usage\n".to_string(),
        })
    );
}

#[rstest]
#[tokio::test]
async fn reads_bundle_reference_text_at_max_size(skill_read_fixture: SkillReadFixture) {
    let response = read_skill_file(&skill_read_fixture.skill, "references/max_size.md").await;

    match response {
        SkillReadFileResponse::Success(success) => {
            assert_eq!(success.skill, "deploy-docs".to_string());
            assert_eq!(success.path, "references/max_size.md".to_string());
            assert_eq!(success.mime_type, "text/markdown".to_string());
            assert_eq!(success.content.len(), MAX_SKILL_READ_FILE_BYTES as usize);
        }
        other => panic!(
            "expected Success for references/max_size.md at MAX_SKILL_READ_FILE_BYTES, got: {:?}",
            other
        ),
    }
}

```

1. Ensure `MAX_SKILL_READ_FILE_BYTES` is in scope in this test module. If it is not already imported (it likely is for the `oversized_text_returns_file_too_large` test), add something like:
   `use crate::skills::file_read::MAX_SKILL_READ_FILE_BYTES;`
   at the top of `src/skills/file_read/tests.rs`.
2. Add a `references/max_size.md` file to the `deploy-docs` test bundle used by `SkillReadFixture`. Its size **must be exactly** `MAX_SKILL_READ_FILE_BYTES` bytes. The content can be arbitrary markdown as long as the total byte length matches the constant.
3. If the test bundle layout is defined via helper code or fixtures elsewhere (e.g., in a builder for `deploy-docs`), update that code to include `references/max_size.md` with the required size instead of creating the file manually.
</issue_to_address>

### Comment 2
<location path="src/skills/file_read.rs" line_range="60" />
<code_context>
+pub struct SkillReadFileSuccess {
+    /// Canonical skill identifier used for the read.
+    pub skill: String,
+    /// Normalised bundle-relative path that was read.
+    pub path: String,
+    /// Best-effort media type inferred from the bundle-relative path.
</code_context>
<issue_to_address>
**issue (review_instructions):** "Normalised" should use -ize spelling per en-GB-oxendict, e.g. "Normalized".

The project style requires en-GB-oxendict spellings, which prefer -ize/-yse forms. Please change "Normalised" here to "Normalized".

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Comments must use en-GB-oxendict (-ize / -yse / -our) spelling and grammar.

</details>
</issue_to_address>

### Comment 3
<location path="src/skills/file_read.rs" line_range="73" />
<code_context>
+pub struct SkillReadFileErrorResponse {
+    /// Canonical skill identifier, or the requested identifier for unknown skills.
+    pub skill: String,
+    /// Requested or normalised bundle-relative path.
+    pub path: String,
+    /// Stable, model-facing error payload.
</code_context>
<issue_to_address>
**issue (review_instructions):** "normalised" should use -ize spelling per en-GB-oxendict, e.g. "normalized".

Same as above: to comply with en-GB-oxendict, please update "normalised" to "normalized" in this comment.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Comments must use en-GB-oxendict (-ize / -yse / -our) spelling and grammar.

</details>
</issue_to_address>

### Comment 4
<location path="docs/execplans/1-3-4-skill-read-file-interface.md" line_range="314" />
<code_context>
+
+## Surprises & discoveries
+
+- Observation: `leta workspace add` succeeded, but the first Rust LSP query
+  failed because `rust-analyzer` was not installed for the active toolchain.
+  Evidence: `leta grep ...` reported that the Rust language server failed to
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “LSP” is used without being expanded on first use; please spell it out (for example, “Language Server Protocol (LSP)”).

Per the style guidance, uncommon acronyms should be defined the first time they appear. Here, “LSP” is introduced without expansion. Consider wording like “Rust Language Server Protocol (LSP) query” on first use.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 5
<location path="docs/execplans/1-3-4-skill-read-file-interface.md" line_range="240" />
<code_context>
+  Mitigation: add attenuation only after unit tests prove no writes, network
+  calls, generic filesystem paths, or state mutation are involved.
+
+- Risk: Python e2e coverage may need deterministic tool-call plumbing rather
+  than a direct unit-level invocation.
+  Severity: medium.
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “e2e” is used without expansion; please define it on first use (e.g. “end-to-end (e2e)”).

The instructions require defining uncommon acronyms when they first appear. “e2e” is introduced here without explanation. Something like “Python end-to-end (e2e) coverage” would satisfy the requirement.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Please investigate the cause of the following issue using codegraph exploration and research, identify a fix and provide an AI coding agent prompt for the fix:

~~~
Run cargo clippy --all --benches --tests --examples --no-default-features --features libsql -- -D warnings
  cargo clippy --all --benches --tests --examples --no-default-features --features libsql -- -D warnings
  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
  env:
    CARGO_HOME: C:\Users\runneradmin\.cargo
    CARGO_INCREMENTAL: 0
    CARGO_TERM_COLOR: always
    CACHE_ON_FAILURE: false
    Updating crates.io index
 Downloading crates ...
...
error: function `metadata_for_opened_file` is never used
   --> src\skills\file_read\io.rs:101:10
    |
101 | async fn metadata_for_opened_file(file: &tokio::fs::File) -> Result<u64, SkillReadFileError> {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D dead-code` implied by `-D warnings`
    = help: to override `-D warnings` add `#[expect(dead_code)]` or `#[allow(dead_code)]`

error: could not compile `ironclaw` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `ironclaw` (lib test) due to 1 previous error
Error: Process completed with exit code 1.
~~~

File: src/skills/file_read/io.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+async fn open_readonly_no_follow(path: &Path) -> Result<tokio::fs::File, SkillReadFileError> {
+    // Non-Unix platforms do not have the Unix O_NOFOLLOW path used above.
+    // This fallback relies on the earlier symlink_metadata rejection and the
+    // later validate_opened_file size check: file size must match stored size
+    // after open or the open is rejected. TODO: use platform-specific atomic
+    // no-follow open semantics for non-Unix targets when this tool supports
+    // those platforms as first-class deployment environments.
+    tokio::fs::File::open(path)
+        .await
+        .map_err(|_| io_error("File is not available for reading"))
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/skills/file_read/io.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 7 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests lack platform-specific guards and don't detect TOCTOU race vulnerability flagged in review comments. | Gate Linux-only tests with #\[cfg(target\_os = "linux")\] to match implementation; cap read\_to\_end at MAX\_SKILL\_READ\_FILE\_BYTES+1 to fail immediately on oversized files. |
| Unit Architecture | ❌ Error | Unbounded read\_to\_end() despite capacity cap (io.rs line 86). Tests lack #\[cfg(target\_os = "linux")\] gates but implementation fails on non-Linux. | Cap reads at MAX\_SKILL\_READ\_FILE\_BYTES+1 bytes during read. Gate max-size and symlink tests with #\[cfg(target\_os = "linux")\] per review. |
| Security And Privacy | ❌ Error | Tests expect successful file reads on all platforms but implementation returns IoError on non-Linux, causing panics. This review finding remains unaddressed. | Add #\[cfg(target\_os = "linux")\] to file\_read/tests.rs for six test functions that expect Success responses on all platforms. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests lack platform-specific gating: reads\_bundle\_reference\_text\_at\_max\_size is ungated and panics on non-Linux. read\_to\_end lacks bounded reads (TOCTOU race). Review comments unaddressed. | Add #\[cfg(target\_os="linux")\] to platform-specific tests. Bound reads with take() to prevent unbounded memory consumption on file growth after stat. |
| Testing (Property / Proof) | ⚠️ Warning | Property tests for path validation present (proptest), but review comments on unbounded reads and platform test gating unadressed; tests lack Linux-only guards. | Add #\[cfg(target\_os = "linux")\] to max-size and symlink tests; cap read\_file\_contents at MAX+1 bytes using Take wrapper for early termination. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Structured JSON response types lack insta snapshot tests recommended for meaningful output expectations; unresolved platform-gating issues in tests. | Add insta snapshot tests for SkillReadFileResponse payloads; gate tests with #\[cfg(target\_os="linux")\] as recommended in review comments. |
| Observability | ⚠️ Warning | New code provides no logging or tracing at decision points (path validation, I/O failures, symlink rejection, platform-specific errors) despite multiple failure modes requiring production diagnostics. | Add structured tracing at validation/error boundaries with stable contextual fields (skill ID, path, error code) using existing patterns in src/skills/ to enable maintainers to diagnose failures. |
| Performance And Resource Use | ⚠️ Warning | read\_to\_end() unbounded allowing file growth past metadata checks to consume unbounded memory. reads\_bundle\_reference\_text\_at\_max\_size fails on non-Linux where io.rs returns IoError not Success. | Add take(MAX+1) to read\_file\_contents. Gate reads\_bundle\_reference\_text\_at\_max\_size to #\[cfg(target\_os = "linux")\]. |
| Concurrency And State | ⚠️ Warning | Test `reads_bundle_reference_text_at_max_size` expects Success on all platforms but implementation returns IoError on non-Linux per review comment requirements. | Add #\[cfg(target\_os = "linux")\] to test or add assertions for non-Linux IoError behaviour per review comments about platform-specific gating. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Review comment not addressed: tests expecting Success (reads\_bundle\_reference\_text\_at\_max\_size, reads\_skill\_entrypoint) lack #\[cfg(target\_os = "linux")\] despite non-Linux failing closed. | Guard tests expecting Success with #\[cfg(target\_os = "linux")\] (non-Linux openat2 unsupported); consider asserting IoError on non-Linux. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for the following concern and provide an AI coding agent prompt for the fix:

**Testing — add `insta` snapshot tests for `SkillReadFileResponse` JSON payloads**  
Cover the serialised shapes of `SkillReadFileResponse::Success`, each `SkillReadFileErrorCode` variant, and the metadata payload to prevent silent serialisation regressions.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for the following concern and provide an AI coding agent prompt for the fix:

**Observability — add structured tracing to `skill_read_file` decision points**  
Add `tracing::debug!`/`tracing::warn!` at path-validation rejection, `openat2` error, size-cap breach, and UTF-8 decode failure with stable fields (`skill_id`, `path`, `error_code`), following existing patterns in `src/skills/`.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/skills/file_read/tests.rs](https://github.com/leynos/axinite/pull/188/files/6d5bbce2a44ef89df889cb2c84908599bac1d79c#diff-91d69f15f1bc2b08bc7ac2b076a2dcb5eb7b706c93ebf998452b74d7e87eec11)

Comment on lines +345 to +356

~~~rust
fn snapshot_error_io_error() {
    let response = SkillReadFileResponse::Error(SkillReadFileErrorResponse {
        skill: "deploy-docs".to_string(),
        path: "references/usage.md".to_string(),
        error: SkillReadFileError {
            code: SkillReadFileErrorCode::IoError,
            message: "File is not available for reading".to_string(),
            metadata: None,
        },
    });
    assert_json_snapshot!("skill_read_file_error_io_error", &response);
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=188&biomarker=Code+Duplication&filename=src%2Fskills%2Ffile_read%2Ftests.rs&method=)
The module contains 5 functions with similar structure: snapshot_error_file_too_large,snapshot_error_invalid_utf8,snapshot_error_io_error,snapshot_error_non_inline_asset and 1 more functions

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/skills/file_read/tests.rs: 

### What lead to degradation?
The module contains 4 functions with similar structure: snapshot_error_file_too_large,snapshot_error_invalid_utf8,snapshot_error_io_error,snapshot_error_non_inline_asset

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/skills/file_read/tests.rs](https://github.com/leynos/axinite/pull/188/files/6d5bbce2a44ef89df889cb2c84908599bac1d79c#diff-91d69f15f1bc2b08bc7ac2b076a2dcb5eb7b706c93ebf998452b74d7e87eec11)

Comment on lines +345 to +356

~~~rust
fn snapshot_error_io_error() {
    let response = SkillReadFileResponse::Error(SkillReadFileErrorResponse {
        skill: "deploy-docs".to_string(),
        path: "references/usage.md".to_string(),
        error: SkillReadFileError {
            code: SkillReadFileErrorCode::IoError,
            message: "File is not available for reading".to_string(),
            metadata: None,
        },
    });
    assert_json_snapshot!("skill_read_file_error_io_error", &response);
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=188&biomarker=Code+Duplication&filename=src%2Fskills%2Ffile_read%2Ftests.rs&method=)
The module contains 5 functions with similar structure: snapshot_error_file_too_large,snapshot_error_invalid_utf8,snapshot_error_io_error,snapshot_error_non_inline_asset and 1 more functions

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/skills/file_read/tests.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: snapshot_error_file_too_large,snapshot_error_non_inline_asset

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please investigate the cause of the following lint violations in the Windows branch of the CI matrix using codegraph exploration and research, identify a fix and provide an AI coding agent prompt for the fix:

~~~

error: unused import: `path_not_readable`
 --> src\skills\file_read\io.rs:7:40
  |
7 | use super::validation::{is_asset_path, path_not_readable};
  |                                        ^^^^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

error: could not compile `ironclaw` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `ironclaw` (lib test) due to 1 previous error
Error: Process completed with exit code 1.
~~~

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Snapshot test uses hardcoded fabricated error message instead of actual validation function output, creating a vacuous test. | Replace hardcoded snapshot test with real read\_skill\_file call to ensure snapshot guards the actual runtime contract, not fabricated values. |
| Module-Level Documentation | ❌ Error | The `file_read` module in `src/skills/mod.rs` lacks a documentation comment, unlike other public modules. Module re-exports require doc comments to guide API consumers. | Add a documentation comment above `pub mod file_read;` in `src/skills/mod.rs` explaining the module purpose and relationship to skill bundle file reads. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Snapshot test hard-codes fabricated message instead of guarding runtime contract; duplicated error tests lack parameterisation despite review comment. | Fix snapshot\_error\_path\_not\_readable to use actual runtime helper; collapse 6 duplicated snapshot tests into single `@rstest` parameterised case. |
| Testing (Property / Proof) | ⚠️ Warning | Critical security invariants (size bounds, symlink rejection) lack property tests; path validation proptest is weak (implication-based, not comprehensive coverage). | Add proptest for MAX\_SKILL\_READ\_FILE\_BYTES; add proptest for symlink rejection; strengthen disallowed\_generated\_paths to directly test all failure cases. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Snapshot test hardcodes error message that differs from the actual runtime helper output, defeating the snapshot's purpose of guarding real contract. | Rewrite snapshot\_error\_path\_not\_readable() to call read\_skill\_file with a traversal path and snapshot the actual runtime response, guaranteeing the snapshot documents real behaviour. |
| Observability | ⚠️ Warning | Missing observability for unknown skill lookup: when a skill does not exist, no tracing is emitted at tool or module level to diagnose the operational failure. | Add tracing::debug! at SkillReadFileTool.execute when lookup\_skill returns None, logging skill name and error code, before creating the unknown\_skill response. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

<html><body>
<h3 dir="auto">❌ Failed checks (1 error, 1 warning)</h3>
<markdown-accessiblity-table data-catalyst="">
Check name | Status | Explanation | Resolution
-- | -- | -- | --
Testing (Overall) | ❌ Error | The reads_skill_entrypoint test only asserts the response is Success(_) without verifying actual content, mime type, or field values. | Replace assert!(matches!(response, Success(_))) with full equality checks on skill, path, mime_type, and content fields like the reads_bundle_reference_text test.
Developer Documentation | ⚠️ Warning | ExecPlan 1.3.4 status is "IN PROGRESS" but Outcomes  declares "Implementation is complete." Status line must match completion  state. | Change line 8 in docs/execplans/1-3-4-skill-read-file-interface.md from "Status: IN PROGRESS" to "Status: COMPLETE".

</markdown-accessiblity-table>
</body>
</html>

**Comment by leynos** [comment]
@coderabbitai review
